Feature/navigation

Frans-Lukas opened the pull request:
<!--
- Add the issue number here.

- If you haven't solved the issue completely use "Linked issue #{issue_number}.
- After solving the issue completely change it to "Fixes #{issue_number}.
-->
Fixes #2112 kind off

<!-- Add here what changes were made in this issue and if possible provide links. -->
The new UI is ready for a PR! It brings a bottom navigation bar and a drawer layout for some general menu items. Please install and test this branch as I am sure there are some newly introduced bugs from all these changes.

<!-- If possible, please add relevant screenshots / GIFs -->

**Screenshots** 
![Screenshot_1596795531](https://user-images.githubusercontent.com/22193232/89636026-43186900-d8a8-11ea-8804-826f7a0d6333.png)
![Screenshot_1596795536](https://user-images.githubusercontent.com/22193232/89636030-44499600-d8a8-11ea-9fec-6da02e5f4041.png)
![Screenshot_1596795541](https://user-images.githubusercontent.com/22193232/89636031-44e22c80-d8a8-11ea-8f9b-de39fb21ee9d.png)

macgills commented on app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt:
```diff
@@ -89,12 +93,14 @@ class ZimManageViewModel @Inject constructor(
 ) : ViewModel() {
   sealed class FileSelectActions {
     data class RequestOpen(val bookOnDisk: BookOnDisk) : FileSelectActions()
```
Shouldn't this be deleted? Or should it just be used instead of `NavigateTo`

I am not against some spring cleaning on this branch where you run the formatter over the codebase and delete any unused classes using the inspections

macgills commented on app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.kt:
```diff
@@ -61,7 +61,7 @@ import org.kiwix.kiwixmobile.zim_manager.library_view.adapter.LibraryListItem
 import javax.inject.Inject
 
-class LibraryFragment : BaseFragment() {
+open class LibraryFragment : BaseFragment() {
```
this doesn't need to be open or really exist at all, should just be OnlineLibraryFragment

macgills commented on core/src/main/java/org/kiwix/kiwixmobile/core/main/ToolbarScrollingKiwixWebView.java:
```diff
@@ -32,15 +32,27 @@
   public ToolbarScrollingKiwixWebView(Context context, WebViewCallback callback, AttributeSet attrs,
     ViewGroup nonVideoView, ViewGroup videoView, CoreWebViewClient webViewClient, View toolbarView,
     View bottomBarView, SharedPreferenceUtil sharedPreferenceUtil) {
     super(context, callback, attrs, nonVideoView, videoView, webViewClient);
```
you should invoke the other constructor with `this` so you don't repeat simple mappings that could introduce subtle bugs. Not very relevant in kotlin because it is an actually decent language

4shutosh commented on core/src/main/res/layout/nav_main.xml:
```diff
@@ -17,18 +16,14 @@
-    app:layout_constraintTop_toTopOf="parent" />
-
-</androidx.constraintlayout.widget.ConstraintLayout>
+    android:layout_height="wrap_content"
```
removing marginTop and using `android:fitsSystemWindows="true"` would be better. 
Using margin may vary outputs on different devices. 
Here's an example :
![Webp net-resizeimage (5)](https://user-images.githubusercontent.com/53831987/89781039-2d1cd980-db30-11ea-8714-f503cc1f5e00.png)
`android:fitsSystemWindows="true"` will work fine here.:smile:

Frans-Lukas replied:
I suspected something would be odd with the margin. Thanks!

4shutosh commented on core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt:
```diff
@@ -42,42 +67,141 @@ abstract class CoreMainActivity : BaseActivity(), WebViewProvider {
+    handleDrawerOnNavigation()
+  }
+
+  private fun openBookmarksActivity() {
```
Opening a fragment needs to be followed by disabling the navigation drawer which you are correctly implementing using `handleDrawerOnNavigation()`. But what if that fragment is not opened from this activity the navDrawer won't be disbaled. 
For example if I bookmark a page from reader and open the bookmarks page using the snackbar action the navigation drawer won't be disabled.
This can be handeled using the `onOptionsItemSelected` on the fragment itself , if I'm not wrong the bookmarks fragment is using the `PageFragment` so making some changes there would fix this issue.

Frans-Lukas replied:
Good catch! Will fix.

s-ayush2903 commented:
The PR was tested with the changes merged in the latest `develop` branch, here are the few observations made:
* The app lags for a whlie(freezes for a msec) navigating from the drawer to `Settings` , `Help` and `Host`
* In the dark mode the color of all the blue buttons need not be changed(i feel so) as the dark mode kinda adds up a whitish layer on the button which doesn't make it look good so the previous one looks better.
* There is also a glitch in `Hosting Books Section`, after turning on the server when one tries turning it off from the _Notifications Panel_, (probably)another another instance of the previously opened section is lauched and on that instance on the clicking the *Stop Server* button although displays the text that _Server has been stopped_ but the button present there isn't updated accordingly, to see the changes/updates one has to navigate back to the (so called previous instance) by clicking the back button or the System navigation and the updated button is displayed. For more clear reference, a gif has been attached:

|Host Fragment|
|--------------------|
|![ss](https://user-images.githubusercontent.com/56556919/89831147-f91ad600-db7a-11ea-8a30-1738376a8b24.gif)|


*  Steps to reproduce:
   * Start the server for hosting the books via nav drawer
   * Now drag down from the notifications panel above the notification that shows the "Kiwix Hotspot", click on it a new screen is         launched.
   * Now click on the _Stop server_ button and notice that althogh the toast is shown and indeed the server is stopped but the text       in the button pressed previously(for starting server) in not updated yet 
    * To see the updated button, one has to navigate back to the previous screen and notice the update

* The app also doesn't has any backstack manipulation on bottom nav frags(like facebook and youtube), if this is intentional then cool but it doesn't look good when app unexpectedly exits when in the reader(or any other fragment) user presses the back button, if that is the case then it may be via this: the app should exit when back button is pressed twice.

* And while using the app i felt that still there exist few memory leaks taking place in ViewModels(for being precise which ones, i'll look into code) that are affecting in a very subtle way that can be seen in profiler or can be felt if tested on a fast device .

Till now that's it, i'll be constantly updating it as I go thorugh the module :)

Frans-Lukas commented:
@s-ayush2903 great feedback, will look into tomorrow. Thanks!

Frans-Lukas commented:
> The app lags for a whlie(freezes for a msec) navigating from the drawer to Settings , Help and Host

Will hopefully be fixed in the settings-to-fragment #2241.

> In the dark mode the color of all the blue buttons need not be changed(i feel so) as the dark mode kinda adds up a whitish layer on the button which doesn't make it look good so the previous one looks better.

Could you clarify what buttons you mean? They might have been part of a contrast audit which changed the colors of a bunch of views to increase readability/contrast. 

> The app also doesn't has any backstack manipulation on bottom nav frags(like facebook and youtube), if this is intentional then cool but it doesn't look good when app unexpectedly exits when in the reader(or any other fragment) user presses the back button, if that is the case then it may be via this: the app should exit when back button is pressed twice.

This was a conscious decision to mimic google drive, wikipedia etc. However, I agree that sometimes it can feel a bit odd pressing back and exiting the app, after having navigated using the bottom nav bar.

> And while using the app i felt that still there exist few memory leaks.

We fixed a bunch of memory leaks in one of the feature/one-activity branches, they should perhaps be directly merged with feature/navigation. 

> One more thing that was observed with the side and bottom nav drawer is that wen one drags and opens the side side nav drawer then at the bottom most part of it, some of its area is clipped by the BottomNavBar.

I don't see this as an issue, just an artifact of using these components. It should appear on any app using a drawer and bottom bar?

Frans-Lukas commented:
> Steps to reproduce:
> * Start the server for hosting the books via nav drawer
> * Now drag down from the notifications panel above the notification that shows the "Kiwix Hotspot", click on it a new screen is launched.
> *Now click on the Stop server button and notice that althogh the toast is shown and indeed the server is stopped but the text in the button pressed previously(for starting server) in not updated yet
> * To see the updated button, one has to navigate back to the previous screen and notice the update

Nice find! This is also an issue on the develop branch. Would you mind creating it as a separate issue @s-ayush2903 ?

s-ayush2903 commented:
Cool, I'll be opening it and will respond if i can figure out its solution

4shutosh commented:
> Could you clarify what buttons you mean? They might have been part of a contrast audit which changed the colors of a bunch of views to increase readability/contrast.
Example the get started button in IntroActivity the blue color needs a change in dark mode. Are we not using colors.xml (night) values anymore to make the same color name use different value in dark mode?

4shutosh commented:
Also, I would like to suggest using a custom class (Adapter) extending a ViewPager or ViewPager2  for the bottom navigation than just using a fragment container it will be easier to add animations, handling the backstack as well as saving the fragment states (cause as of now we everything gets refreshesd when we change a destination )

Frans-Lukas commented:
> 
> 
> > Could you clarify what buttons you mean? They might have been part of a contrast audit which changed the colors of a bunch of views to increase readability/contrast.
> 
> Example the get started button in IntroActivity the blue color needs a change in dark mode. Are we not using colors.xml (night) values anymore to make the same color name use different value in dark mode?

Ah! We changed to use themes instead of overriding colors https://github.com/kiwix/kiwix-android/pull/2122 here. In the contrast audit I changed the primary color to contrast correctly on the dark background as the old primary color had poor contrast to dark backgrounds. Examples include the button text in the "External link" dialog, or the preference titles in the settings activity. I'll look into changing the button style or the type of text that uses the primary color as text color.

Frans-Lukas commented:
> 
> 
> Also, I would like to suggest using a custom class (Adapter) extending a ViewPager or ViewPager2 for the bottom navigation than just using a fragment container it will be easier to add animations, handling the backstack as well as saving the fragment states (cause as of now we everything gets refreshesd when we change a destination )

In my opinion this would require quite a re-write and might better fit a new ticket (because the bottom nav bar works (mostly)), instead of solving bugs introduced by this branch.

Frans-Lukas commented:
@s-ayush2903 The zim host issue might be solved by converting it to a fragment, which is done in #2251 . So you might not need to work on this issue right now, I'll keep you updated!

macgills commented:
>Also, I would like to suggest using a custom class (Adapter) extending a ViewPager or ViewPager2 for the bottom navigation than just using a fragment container it will be easier to add animations, handling the backstack as well as saving the fragment states (cause as of now we everything gets refreshesd when we change a destination )

As per this [answer](https://stackoverflow.com/a/52146506/12225251) there is a workaround for backstacks but also it is being actively developed by the Navigation team and will become a real feature hopefully sooner rather than later as 2.3.0 was 24th June and next release should be 2.4.0-alpha01 which was scheduled to have the feature.

4shutosh commented:
Yess, multiple back stack support will simplify things. Thank you for sharing.
But we'll still need to save the fragments states (recyclerViews, etc) manually.

s-ayush2903 commented:
>  Could you clarify what buttons you mean? They might have been part of a contrast audit which changed the colors of a bunch of views to increase readability/contrast.

I meant only the buttons that are **blue** in color, as the night mode adds some whitish layer on the top of them, I think it would be better if the buttons in blue don't get their color changed even in the night mode(it's just my very personal opinion). For more clarity let's have a look below

|In Light Mode|Dark Mode is turned On|
|-------------------|----------|
|![Screenshot_20200813-114352_Kiwix](https://user-images.githubusercontent.com/56556919/90100755-f7a80400-dd5a-11ea-9a68-34529b1958ef.jpg)|![Screenshot_20200813-114314_Kiwix](https://user-images.githubusercontent.com/56556919/90100780-0393c600-dd5b-11ea-99e8-c7e1ea3b2eab.jpg)|

macgills commented:
```
e: /Users/runner/work/kiwix-android/kiwix-android/app/src/test/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModelTest.kt: (59, 79): Unresolved reference: RequestOpen
> Task :app:compileDebugUnitTestKotlin
e: /Users/runner/work/kiwix-android/kiwix-android/app/src/test/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModelTest.kt: (426, 51): Unresolved reference: RequestOpen
e: /Users/runner/work/kiwix-android/kiwix-android/app/src/test/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModelTest.kt: (427, 23): Unresolved reference: OpenFile
```

unit tests failing to compile

macgills commented on app/src/main/java/org/kiwix/kiwixmobile/main/KiwixMainActivity.kt:
```diff
@@ -18,24 +18,129 @@
+    bottom_nav_view.setupWithNavController(navController)
+  }
+
+  override fun onPostCreate(savedInstanceState: Bundle?) {
```
postCreate

macgills commented on custom/src/main/java/org/kiwix/kiwixmobile/custom/main/CustomMainActivity.kt:
```diff
@@ -19,25 +19,83 @@
+  }
+
+  override fun onPostCreate(savedInstanceState: Bundle?) {
+    super.onPostCreate(savedInstanceState)
```
postCreate, also this code looks like it fits in Core

Frans-Lukas replied:
This would have to be called after the creation lifecycle step in Core? I'd agree with this answer https://stackoverflow.com/a/16564121 that it's ok to use in a parent activity.

macgills replied:
![image](https://user-images.githubusercontent.com/3358282/90246191-f1ce2380-de2b-11ea-91b3-881de05d6bee.png)

So call it in onStart if it happens after `OnCreate` or just at the end of OnCreate. I truly hate this method as `postCreate` is not what happens immediately after `onCreate`

Frans-Lukas replied:
Oof, it's a lie

Frans-Lukas replied:
Hmm or I guess everything that happens after oncreate could be called postOnCreate

macgills replied:
yeah but it violates the principle of Least Surprise. I think the typical assumption is `this happens immediately after create` and you have to read the docs to discover it is a bastard.

Every android developer has to know the activity lifecycle, if only for it being asked in every interview, so lean on that shared knowledge

s-ayush2903 commented:
An update on testing:
This time _CustomApp_ was taken on testing.
The custom app seems to be jammed completely.
* It crashes on its first install(please reinstall the app after the uninstallation as previous settings may not lead it to crash) with a border of red lines over the screen showing that there was an attempt made to launch the app and it disappears after the crash occurs leaving user with the permission request for Media access.
* After this when app is next launched(not on reinstallation) it says: (umm..., let me attach the exact visuals) [Here they are :link: ](https://drive.google.com/file/d/18aXlGGoN76Ox2EsJWbT9B-pBnkWobVDf/view?usp=sharing)

Okay, i know all it needs is a subtle change to fix the out of indices jsonError, so i'm opening a PR for the same fix

Frans-Lukas commented:
@s-ayush2903 great find! Go ahead and create the ticket and PR 👌

s-ayush2903 commented:
#2285 Opened

macgills commented:
I am happy to merge this tomorrow for a beta release only. Has been a while coming and user feedback would be invaluable.

I think the only outstanding issue is custom apps and we can direct focus at that before the actual release.

4shutosh commented:
Yayy! 🌟

Frans-Lukas commented:
yayy indeed!!! 🥳
